### DISCUSSION TITLE
Need cockpit.spawn() to wait till the command (bash script with a couple of arguments) provided completes its execution entirely

### DISCUSSION THREAD

**Discussion by sjagtap** [cockpit-project/cockpit#19606] [Q&A] [answered]
Here's what I am trying to accomplish:
1- I want to launch VMs using qcow2 images. The following two options "Import and Run" & "Import and Edit" are provided to do the same:
<img width="629" alt="image" src="https://github.com/cockpit-project/cockpit/assets/2447496/3dd63789-2c3e-4845-bf24-743d2ef77c50">
In the disk image section, an option to specify either a .qcow2 OR a zipped version of the same .qcow2.gz is to be provided.
My solution works seamlessly when a .qcow2 image is specified. However, I am facing an issue when specifying a .qcow2.gz image

Root cause of the above issue when trying to deploy a .qcow2.gz image:
1- In the background, the .qcow2.gz file needs to be unzipped (I am using the gunzip command to accomplish the same) and copied to "/var/lib/libvirt". This path is then specified in "const vmParams" of cockpit's "createVmDialog.jsx" for VM creation
2- I have written the "decompression and copying to destination " code in a bash script which I am executing using cockpit.spawn() as seen below:
```
async doReplicaOfImage(targetPath, vmName, diskSizeInGb) {
    const { onAddErrorNotification } = this.props;
    try {
        await cockpit.spawn(
            ["/usr/bin/manage_image.sh", this.state.source, targetPath, diskSizeInGb],
            { err: "out" }
        );
        return 0;
    } catch (exception) {
        console.error(`spawn 'vm creation' returned error: "${JSON.stringify(exception)}"`);
        onAddErrorNotification({
            text: cockpit.format(_("Creation of VM $0 failed, --$1, --$2, --$3"), vmName, this.state.source, targetPath, diskSizeInGb),
            detail: exception.message.split(/Traceback(.+)/)[0],
        });
    }
}
```
Inside the manage_image.sh script, I am performing a "/bin/gunzip -c "$source_path" > "$dest_path" operation followed by the validation of the gunzipped image thus created at dest_path.
3- The qcow2.gz image is very large (>6gb) and hence requires a certain amount of time to get decompressed. So when the "Import and Run" option is used, it's not getting sufficient time to decompress.
As you can see in the code snippet I pasted above, I have used the async and await keywords to try and wait for the script to finish execution, but cockpit.spawn() doesn't seem to be helping me accomplish it. As per the cockpit documentation, "The spawned process is a [promise] that will complete if the process exits successfully, or fail if there's a problem."
What I could understand from the above was that the script will run in the background and the createVmDialog.jsx code will continue to execute while the script finishes execution. This leads to the problem wherein the gunzipped image has not yet been fully decompressed at the destination path, but VM creation process begins (eventually failing)

TLDR: How to ensure that the createVmDialog.jsx code waits for VM creation, until the execution of my bash script (which decompresses a .qcow2.gz image and validates the decompressed image at the target location) completes fully?
cockpit.spawn() along with async and await isn't accomplishing it for me.

**Comment by martinpitt** [comment, marked as answer]
> What I could understand from the above was that the script will run in the background and the createVmDialog.jsx code will continue to execute while the script finishes execution

No, this isn't what happens. `cockpit.spawn().then(() => console.log("done"))`, or alternatively, `await cockpit.spawn()` really does wait until the process finishes. You can save the result of spawn into a variable, and add some console.log() messages to follow when it finishes or catches.

Two theories: (1) you don't  `await` the call of `doReplicaOfImage()`? (2)  Your manage_image.sh script is failing early? I suggest adding `set -x` and writing it to the console.

**Reply by sjagtap** [reply]
Thanks @martinpitt , your first theory was correct.
I added an await before the call to doReplicaOfImage() and it worked like a charm.

Thank you for your prompt response! And let me use this opportunity to thank you for your responses to other posts also, which have been helpful as well.

**Comment by martinpitt** [comment]
Very nice, cheers!
